SAM4L analog comparator driver

dverhaert opened the pull request:
### Pull Request Overview

This pull request adds a new capsule: the Analog Comparator. It controls a number of Analog Comparators (ACs), which compares voltages and give an output depending on this comparison. I've also written an implementation for the Cortex M4, a HIL, a userland application capable of doing basic operation, a component interface and a README in doc/syscalls.

### Testing Strategy

This pull request was tested extensively by me on the Hail; on there it works as expected. 

### TODO or Help Wanted

- [ ]  Testing on imix

### Documentation Updated

- [x] Kernel: Updated the relevant files in `/docs`, or no updates are required.
- [x] Userland: Added/updated the application README, if needed.

### Formatting

- [x] Ran `make formatall`.

ppannuto commented on capsules/src/acifc.rs:
```diff
@@ -0,0 +1,61 @@
+//! Provides userspace access to the ACIFC interface.
+//!
+// !## Instantiation
```
`// !##` -> `//! ##`

ppannuto commented on chips/sam4l/src/acifc.rs:
```diff
@@ -0,0 +1,370 @@
+
+register_bitfields![u32,
+	Control [
+		// Analog comparator test mode. Equal to 1 means AC outputs will be bypassed with values in AC test register.
```
If you make these doc comments (`///` instead of `//`) they'll show up in the auto-generated documentation

ppannuto commented on chips/sam4l/src/acifc.rs:
```diff
@@ -0,0 +1,370 @@
+    }
+
+    fn enable_clock(&self) {
+        unsafe {
```
If you rebase on latest master (or merge it in), calls to pm shouldn't require `unsafe` anymore

ppannuto commented on chips/sam4l/src/acifc.rs:
```diff
@@ -0,0 +1,370 @@
+	]
+];
+
+const BASE_ADDRESS: *mut AcifcRegisters = 0x40040000 as *mut AcifcRegisters;
```
Add a line here like:
`const ACIFC_REGS: StaticRef<AcifcRegisters> = unsafe { StaticRef::new(BASE_ADDRESS as *const AcifcRegisters) };`

Then everywhere below you can just use ACICF_REGS without unsafe

ppannuto commented on chips/sam4l/src/acifc.rs:
```diff
@@ -0,0 +1,370 @@
+    }
+
+    // Handling interrupts not yet implemented.
+    pub fn handle_interrupt(&mut self) {}
```
You can use rust's `unimplemented!("ACIFC Interrupt");` in the function body here (will print nice error if it ever gets called somehow)

niklasad1 commented:
Hey, thanks for this PR

If you run ```$ make formatall``` (make sure you have uncrustify installed) then I think Travis will pass and I would appreciate if you could fix the compiler warnings.

dverhaert commented:
Hey,

Thanks for all the valuable feedback! I've implemented all of your requested changes and confirmed that the board still works, so please let me know if there are other things you spot that should be edited/changed, especially regarding the pin numbers and how to resolve that.

Also, slightly unrelated: did compiling change somehow in the last month? I'm asking because I edit my code in Windows (blasphemy, I know) and use a virtual machine with a shared folder to compile/run the code after editing it. However, since merging in the latest master, I get an error:
`error: failed to build archive: Text file busy`
I've done some research, and I believe the problem lies in the way that gedit/glib saves a file by writing to a temporary file, and consequently renaming to the final name. It's a bug which has existed since 2009 and nobody is willing to fix it. Currently the only solution I found after some deep digging is to edit the source code of glib: clearly this is not a clean solution. Could it be that in a recent pull request glib is utilized? If so, could this perhaps be circumvented somehow? That would make my, and possible other Unix noobs', life a lot easier!

alevy commented on capsules/src/acifc.rs:
```diff
@@ -0,0 +1,61 @@
+//! Provides userspace access to the ACIFC interface.
```
`ACIFC` is a pretty SAM4L-specific name. For platform-agnostic components, like the system-call capsule, we should use names that are more meaningful from an application perspective. Perhaps: "AnalogComparator"

dverhaert replied:
I've thought about this for a bit, but reckoned "AC/ACs" was too short to be clear and "AnalogComparators" was too long and might make the code unnecessarily large. How about I change the names at the top to "Analog Comparators", and keep the use of "Acifc" in the code the same?

alevy commented on capsules/src/acifc.rs:
```diff
@@ -0,0 +1,61 @@
+//! Author: Danilo Verhaert <verhaert@cs.stanford.edu>
+
+/// Syscall driver number.
+pub const DRIVER_NUM: usize = 0x07;
```
If we're adding a cross-platform system call driver, should update `doc/syscalls/README.md` and document the interface in `doc/syscalls/00007_analog_compartor.md`.

dverhaert replied:
Done!

alevy commented on kernel/src/hil/acifc.rs:
```diff
@@ -0,0 +1,40 @@
+
+use returncode::ReturnCode;
+
+/// Enum for selecting which edge to trigger interrupts on.
```
Left over commented out code? If so, just remove it.

dverhaert replied:
I left this commented out code for now, in case someone wanted to implement an interrupt option. Thought this might be useful for e.g. zero-crossing, but assumed the main use would be simple voltage comparison, so I focused on some basic implementation for that first. You think I should just remove these comments (and leave the implementation for the future)?

alevy commented on kernel/src/hil/acifc.rs:
```diff
@@ -0,0 +1,40 @@
+
+pub trait Acifc {
+    /// Initialize the ACIFC
+    fn initialize_acifc(&self) -> ReturnCode;
```
Is there a reason this needs to happen explicitly? Is there _always_ an initialization step for analog comparator hardware? If so, what is it meant to do? What are the expectations the caller can have after calling this method?

dverhaert replied:
Two steps are always necessary: Enabling the clock, and enabling the ACIFC by writing a 1 to the Enable register. Do you agree that calling an initialization function is better than always calling the clock and writing to this register?

alevy commented on kernel/src/hil/acifc.rs:
```diff
@@ -0,0 +1,40 @@
+    fn initialize_acifc(&self) -> ReturnCode;
+
+    /// Enable the clock
+    fn enable_clock(&self);
```
Genreally clock management doesn't belong directly in the cross-platform HIL. If it _does_ belong here in this particular case, the comment should explain what it's purpose is exactly.

alevy commented on kernel/src/hil/acifc.rs:
```diff
@@ -0,0 +1,40 @@
+    fn disable_clock(&self);
+
+    /// Do a comparison of two inputs. Output will be 1 when Vinp>Vinn (Vin positive > Vin negative)
+    fn normal_comparison(&self, usize) -> u32;
```
Maybe a more informative name than `normal_`... specifically this method returns whether Vinp > Vinn, so the name should just say that. What about just "compare"?

What is the `usize` argument?

What about a more specific return type? Either a `bool` (if the return value is really only going to be `1` or `!1`. Alternatively, an enum that values only in the possible range.

dverhaert replied:
Returning as bool now. However, when using ReturnCode::SuccessWithValue a usize is expected, so in the capsule it's converted to a usize anyways.

alevy commented on kernel/src/hil/acifc.rs:
```diff
@@ -0,0 +1,40 @@
+    fn normal_comparison(&self, usize) -> u32;
+
+    /// Do a comparison of three inputs. Output will be 1 when Vacbn_x+1 < Vcommon < Vacap_x!
+    fn window_comparison(&self, usize) -> u32;
```
I don't understand from the comments + type what this does :(

alevy commented on kernel/src/hil/acifc.rs:
```diff
@@ -0,0 +1,40 @@
+    fn window_comparison(&self, usize) -> u32;
+
+    /// Do a basic test to ensure everything is working
+    fn test_output(&self) -> ReturnCode;
```
I think this isn't worth including as part of the HIL.

dverhaert replied:
I thought it'd be safe to keep this in for now, since I expected I had to make quite some changes. Will remove it for the final merge

alevy commented on kernel/src/hil/acifc.rs:
```diff
@@ -0,0 +1,40 @@
+    fn test_output(&self) -> ReturnCode;
+}
+
+// pub trait Client {
```
Left over commented out code

dverhaert replied:
Same story as for the interrupt code. Keep it for future implementation or remove it for now?

alevy commented on kernel/src/hil/acifc.rs:
```diff
@@ -0,0 +1,40 @@
+//     EitherEdge,
+// }
+
+pub trait Acifc {
```
High-level, this HIL needs to be fleshed out significantly. To begin with, it would be good to have much more comments. For one, I think trying to actually explain in prose what the purpose of the HIL is (the kind of hardware it abstracts), as well as what users can expect and what implementors should guarantee would help draw out issues with the interface and second, we just need documentation :)

dverhaert replied:
I am always a fan of more documentation. I'll add way more.

dverhaert replied:
What should be the central location of the comments? In sam4l, capsules or the HIL?

dverhaert commented:
Made some more changes, thanks for all the valuable feedback all! How do you feel about the documentation in hil/readme now? Also, anyone know why Travis keeps failing?

For now I've commented out the test function, in the final commit I'll delete it.

alevy commented:
@dverhaert Sorry to leave this hanging for a while...

I'm not versed in using analog comparators, but after prodding @ppannuto, @brghena, and @bradjc, it actually seem like this _might_ be kind of the mirror of the API that makes most sense to focus on. In particular, it seems like we probably want an interrupt-drive API, rather than synchronous sampling (though maybe it's OK to have synchronous sampling as well, cause why not).

In particular, it seems that the applications where it makes sense to use an analog comparator instead of (or in addition to) the ADC are _exactly_ interrupt driven ones: where you want a lower-power peripheral that is waiting for a particular line to pass a certain threshold. Some examples that came up:

  * Get a notification when the voltage on a capacitor that is the energy store for an energy harvesting system drops below a threshold (meaning it's about to run out)

  * A "Clapper" that waits for the noise levels on a microphone are very high but is otherwise totally asleep.

  * Detecting light levels dropping/rising above/below a threshold

Often using the comparator would trigger a more expensive operation on the ADC.

dverhaert commented on kernel/src/hil/analog_comparator.rs:
```diff
@@ -0,0 +1,33 @@
+    /// Do a comparison of two inputs. Output will be 1 when one is higher than
+    /// the other, and 0 otherwise. Specifically, the output is 1 when Vinp>Vinn
+    /// (Vin positive > Vin negative), and 0 if Vinp < Vinn.
+    fn comparison(&self, usize) -> bool;
```
The usize value is chosen to select the analog comparator you want to use, this differs per MCU.

dverhaert commented on kernel/src/hil/analog_comparator.rs:
```diff
@@ -0,0 +1,33 @@
+    /// the output will be 1; it will be 0 if it is outside of the window.
+    /// Specifically, the output will be 1 when Vn_x+1 < Vcommon < Vp_x, and 0
+    /// if Vcommon < Vn_x+1 or Vcommon > Vp_x.
+    fn window_comparison(&self, usize) -> bool;
```
This number represents the choice of the comparators. For example, in the imix choosing windowed mode 1 would mean using comparators 1 and 2. Using windowed mode 2 would mean using comparators 3 and 4.

dverhaert commented on doc/syscalls/00007_analog_comparator.md:
```diff
@@ -0,0 +1,140 @@
+gives a compare output depending on this comparison.
+
+A specific AC is referred to as ACx where x is any number from 0 to n and n is
+the index of last AC module.
```
This was actually something that Brad accidentally copied away. It's fixed now.

dverhaert commented on kernel/src/hil/analog_comparator.rs:
```diff
@@ -0,0 +1,33 @@
+
+use returncode::ReturnCode;
+
+pub trait AnalogComparator {
```
I'm actually working on this right now, planning to have it finished by the end of the week

dverhaert commented on capsules/src/analog_comparator.rs:
```diff
@@ -0,0 +1,60 @@
+    ///        the ACIFC itself.
+    /// - `2`: Perform a simple comparison.
+    ///        Input x chooses the desired comparator ACx (0 or 1 for hail,
+    ///        0-3 for imix)
```
It is defined in the readme in doc/syscalls. But, for clarification, I've added some more explanation at the top of the file

dverhaert commented on kernel/src/hil/analog_comparator.rs:
```diff
@@ -0,0 +1,33 @@
+    /// defined by the positive input of ACx and the negative input of ACx+1,
+    /// the output will be 1; it will be 0 if it is outside of the window.
+    /// Specifically, the output will be 1 when Vn_x+1 < Vcommon < Vp_x, and 0
+    /// if Vcommon < Vn_x+1 or Vcommon > Vp_x.
```
Do you know how to neatly indent triple slash comments? Right now it's all pretty messed up

dverhaert commented on userland/examples/tests/acifc/main.c:
```diff
@@ -0,0 +1,80 @@
+  // Choose your comparator. AC = 0 corresponds to PA06 and PA07, whereas ac = 1
+  // corresponds to PB02 and PB03. On the hail these are the pins DAC and WKP,
+  // and AC2 and AC3 respectively.
+  uint8_t ac = 1;
```
Actually, I don't really see why this isn't general enough. The ADC has this channel-structure because it loops through all of the channels sequentially in userland, whereas for an Analog Comparator you generally only need to use one or two at the same time (which most chips have for sure). Right now, this "index" is the number you select in userland. What would be the merit of adding another structure in the capsule to translate to a channel type? That the code would throw an out of bounds error when someone tries to access an index outside of the accepted range?

brghena replied:
So the channel structure isn't just to allow checking on the selections from userland. The additional feature they provide is limiting capsules to only using channel selections given to them from a board `main.rs` file.

For example, a hypothetical battery-sensing capsule would want to be hooked up to the battery voltage and compare that against some reference voltage. The pins/channels to make that configuration work have to be set by the board `main.rs` file since they will be specific to each design. So from the battery-sensing capsule's perspective, it's given an opaque handle that signifies "the comparator configuration that senses the battery". Without being given this handle, it's never going to be possible to interchangeably function on different platforms.

From this perspective, it's irrelevant whether that handle is a number or a type on the inside. What is important though, is that the handle be some object which can only be created by the board file and only be consumed by the chip driver. Using a `usize` as the structure would hypothetically allow the capsule to craft its own handles, which it is not in a position to do (since the capsule can't know the configuration for every board).

dverhaert commented:
I've been working on developing an interrupt-based implementation for the analog comparators, and thought I had managed to do so. However, I found out that my implementation does a few very peculiar things which I'm unable to put my finger on:
- It prints the opposite of what you would expect: it receives an interrupt when the negative input is higher than the positive input (Vp < Vn), whilst this should happen when Vp > Vn.
- This 'succesful' interrupt reception only works when you start in a situation where Vp > Vn (as in, you connect Vn to ground and Vp to 3v3). If you start listening to tockloader when Vp > Vn (as in, you leave both pins floating or you connect Vp to ground and Vp to 3v3), it doesn't work, and sometimes this goes as far as the kernel panicking.

I'd appreciate it if someone could look into my code, to see what I'm doing wrong and if I'm getting interrupts to userland in the right way. Because there is no standardized way of doing this, I just searched the entire repository for the word "callback" and improvised somewhat on my implementation. The message shown during the kernel panic gives away that my implementation is far from perfect:

`"called `Option::unwrap()` on a `None` value"`

Specifically, I'm not sure about the code handling interrupts on **line 278 of sam4l/acifc.rs** and subscribing/firing interrupts from **line 89 of capsules/analog_comparator.rs.**


Also, I have one other problem I'd like some help with. In userland, I select which analog comparator to use by passing a variable to the kernel. Let's say I want to read out the value of the ACCS register depending on which variable I pass to the kernel. If I pass a 0, I want to read ACCS0, and if I pass a 1, I want to read ACCS1. Currently, I resolve this situation by the following lines of code:

if ac == 0 { result = regs.sr.is_set(Status::ACCS0);} 
else if ac == 1 {  result = regs.sr.is_set(Status::ACCS1);

which is clearly not optimal, especially if more ACs or register commands are involved. Would anyone know a better way to do this?

Thanks guys, all the help is really appreciated 😄

bradjc commented:
The SAM4L is a peculiar beast. Getting the HW to work the way you expect can be tricky. Have you cross checked your code with the ASF?

As for the unwrap, you should use `map` instead which won't panic.

And a big if else or `match` statement seems fine. Working code is better than getting the ideal interface done on the first try.

niklasad1 commented on capsules/src/analog_comparator.rs:
```diff
@@ -0,0 +1,113 @@
+    for AnalogComparator<'a, A>
+{
+    fn fired(&self) {
+        self.callback.get().unwrap().schedule(0, 0, 0);
```
```rust
self.callback.get().map_or_else(|| {
   debug!("callback is None");
   false
}, |mut cb| cb.schedule(0,0,0))
```

bradjc commented:
Closing without prejudice, please re-open when ready for another review.

dverhaert commented:
Ready for another review!

I added Niklas's code, and did some more debugging. There's some good and bad news.

The good news: I finally found out what the problem is, and I can get the analog comparator to work without a hitch!

The bad news: It only works under specific conditions, and I don't understand why.  

So the problem is that **when I hold my fingers near any of the DAC, WKP, RX or TX pins, the kernel somehow receives a ton of interrupts, correspondingly fires a lot of times, causing a lot of messages to be printed and often crashes**. It _should_ only print these messages when the positive input voltage Vinp is higher than the negative input voltage Vinn. 

Only when I'm really careful in connecting the pins to Hail and making sure my fingers are nowhere near the board does everything work as expected. 

I thought this was caused by the DAC and WKP pins on hail being connected to analog comparator 0, whereas I use analog comparator 1 for my tests. However, even when I disable comparator 0 or even change the multiplexing of the pinout this problem occurs.

Does anyone know what could cause this and if this is critical to fix for the merge with master? I'm guessing my fingers cause static electricity to flow in some of the pins, but that should not matter at all for this test (since I use pins A2 and A3). Took me a while to find out what the problem was, really did not expect this.

If you have questions about the code, please let me know and I'll respond asap.

dverhaert commented on chips/sam4l/src/acifc.rs:
```diff
@@ -0,0 +1,392 @@
+
+const BASE_ADDRESS: *mut AcifcRegisters = 0x40040000 as *mut AcifcRegisters;
+const ACIFC_REGS: StaticRef<AcifcRegisters> =
+    unsafe { StaticRef::new(BASE_ADDRESS as *const AcifcRegisters) };
```
I did it this way to match all of the other SAM4L peripheral implementations. Do you think we should change those too, or keep it this way for consistency?

dverhaert commented on kernel/src/hil/analog_comparator.rs:
```diff
@@ -0,0 +1,45 @@
+    /// allowing a measurement on an AC to be made quickly after a measurement
+    /// is triggered, without waiting for the AC startup time. The drawback is
+    /// that when the AC is always on the power dissipation will be higher.
+    fn enable(&self) -> ReturnCode;
```
Enable currently does three things:
- Enable the clock
- Set a bit in control register to activate the Analog Comparator Interface
- Set the ACs in continuous measurement & always on modes

Here, again I followed the example of the other modules (mostly I looked at AES, CRCCU, and GPIO (notice the Oxford comma ;), which all have this). I've added a disable function, and took the opportunity to do the checking for valid ac/window values to the kernel. If they are incorrect, I print a debug and disable/panic the analog comparator.

dverhaert replied:
Also, I'll rebase in the morning but other than that please take a look at the changes

bradjc commented:
I squashed the commits and reverted the changes to Hail (I think we want to leave Hail with the DAC since that is how the pin is labeled).

I'm still not entirely sure how the HIL interface is intended to be used. Looks like a call to `enable()` is first, but then its not clear which functions are blocking, or if calling `enable_interrupts()` is enough to get comparator events or not. Also there is no `disable()` so it is not clear how a user would stop getting interrupts.

dverhaert commented:
I've tried to think about the hil from a platform-independent perspective, removing `enable()` and adding `disable_interrupts()` to the hil. Now, `enable()` is called in the chip-specific code (following #1025) and `disable_interrupts()` can be used to disable interrupts. This can be used for example in the case that a user would like to disable the interrupts for a certain amount of time after receiving one.

Also, I've partially rewritten some of the changes @bradjc reverted. I feel you're right about leaving the DAC and WKP pins for those respective functions since they're labeled, but think A2 and A3 could be used for the analog comparator, couldn't they? 

I've also added some documentation and finished some rough edges, and honestly think the analog comparator is in a pretty good state now, nearly ready to be merged! Please check thoroughly now and let me know if anything is still missing.

bradjc commented on kernel/src/hil/analog_comparator.rs:
```diff
@@ -0,0 +1,44 @@
+    /// will be True (1) when one is higher than the other, and False (0)
+    /// otherwise.  Specifically, the output is True when Vp > Vn (Vin positive
+    /// > Vin negative), and False if Vp < Vn.
+    fn comparison(&self, usize) -> bool;
```
What is the `usize` argument to all of these functions? Presumably the analog comparator channel? I think it would be better to do something like what the ADC HIL does:

https://github.com/tock/tock/blob/f05997c2fdf1c2761a63efc5c612bef968ad6d2e/kernel/src/hil/adc.rs#L7-L10

where the type is specified by the specific implementation.

dverhaert commented:
I finally had some time to look over this again! I added channels, as requested by @bradjc and @brghena as is done in the ADC. Also, I cleaned up the implementation, it looks a lot cleaner now.

One final design decision I'm struggling with is window comparisons. I think that every chip with an analog comparator can do at least a one-off comparison and interrupt comparisons, but I'm not so sure about window comparisons. Also, since the functionality of window comparisons is not very useful directly I am considering removing window comparisons completely: if it's not general, it shouldn't be in the HIL and capsule. What do you guys think about this?

Let's get this merged this week! 🥇

bradjc commented on boards/hail/src/main.rs:
```diff
@@ -156,8 +162,12 @@ unsafe fn set_pin_primary_functions() {
+    PB[03].configure(Some(E)); // A3 - ACIFC ACBP0
+                               // // ADC Mode
+                               // PB[02].configure(Some(A)); // A2 - ADC3
+                               // PB[03].configure(Some(A)); // A3 - ADC4
```
Please revert the changes to Hail so we can leave the default operation of those pins as ADC pins (which is what they are labeled as on the silkscreen).

bradjc commented on capsules/src/analog_comparator.rs:
````diff
@@ -0,0 +1,175 @@
+//! ```rust let acifc = static_init!(
+//! capsules::analog_comparator::AnalogComparator<'static, sam4l::acifc::Acifc>,
+//! capsules::analog_comparator::AnalogComparator::new(&mut
+//! sam4l::acifc::ACIFC)); ```
````
Need newlines after and before the ` ``` `.

bradjc commented on capsules/src/analog_comparator.rs:
````diff
@@ -0,0 +1,175 @@
+//!
+//! ```rust let acifc = static_init!(
+//! capsules::analog_comparator::AnalogComparator<'static, sam4l::acifc::Acifc>,
+//! capsules::analog_comparator::AnalogComparator::new(&mut
````
`new` takes two arguments.

bradjc commented on kernel/src/hil/analog_comparator.rs:
```diff
@@ -0,0 +1,47 @@
+    /// it lies outside of the window.  Specifically, the output will be True
+    /// when Vn_x+1 < Vcommon < Vp_x, and False if Vcommon < Vn_x+1 or Vcommon >
+    /// Vp_x.
+    fn window_comparison(&self, usize) -> bool;
```
What is the `usize` argument?

bradjc commented on kernel/src/hil/analog_comparator.rs:
```diff
@@ -0,0 +1,47 @@
+
+pub trait Client {
+    /// Called when an interrupt occurs.
+    fn fired(&self);
```
How does the client know which channel generated the interrupt?

bradjc commented on kernel/src/hil/analog_comparator.rs:
```diff
@@ -0,0 +1,47 @@
+
+    /// Enable interrupt-based comparison for the chosen AC (e.g. AC1). This
+    /// will make it listen and send an interrupt as soon as Vp > Vn.
+    fn enable_interrupts(&self, channel: &Self::Channel) -> ReturnCode;
```
Can this be called something like `start_comparing()`?

bradjc commented on kernel/src/hil/analog_comparator.rs:
```diff
@@ -0,0 +1,47 @@
+    fn enable_interrupts(&self, channel: &Self::Channel) -> ReturnCode;
+
+    /// Disable interrupt-based comparison for the chosen AC.
+    fn disable_interrupts(&self, channel: &Self::Channel) -> ReturnCode;
```
Then a matching `stop_comparing()` (or similar).

ppannuto commented:
Regarding windowing, I agree it’s specialized. We have some predent with
UART (not that it’s the model HIL..) for a split HIL, with the common basic
operations in the main HIL and then some optional “advanced” interfaces. I
think that applies cleanly here as well.

I’d be inclined to just focus on the primary HIL here and leave the
windowing for another PR

On Tue, Aug 7, 2018 at 00:31 Danilo <notifications@github.com> wrote:

> I finally had some time to look over this again! I added channels, as
> requested by @bradjc <https://github.com/bradjc> and @brghena
> <https://github.com/brghena> as is done in the ADC. Also, I cleaned up
> the implementation, it looks a lot cleaner now.
>
> One final design decision I'm struggling with is window comparisons. I
> think that every chip with an analog comparator can do at least a one-off
> comparison and interrupt comparisons, but I'm not so sure about window
> comparisons. Also, since the functionality of window comparisons is not
> very useful directly I am considering removing window comparisons
> completely: if it's not general, it shouldn't be in the HIL and capsule.
> What do you guys think about this?
>
> Let's get this merged this week! 🥇
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tock/tock/pull/837#issuecomment-410962324>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAUt3scy3z0OYFj7Kdg9ZM74IIdVVjuxks5uOUJfgaJpZM4S4RpO>
> .
>

dverhaert commented:
Implemented changes, please rereview!

bradjc commented on capsules/src/analog_comparator.rs:
```diff
@@ -0,0 +1,156 @@
+    // Start comparing on a channel
+    fn start_comparing(&self, channel: usize) -> ReturnCode {
+        if channel >= self.channels.len() {
+            panic!("Please select a channel which exists on the current board/chip");
```
We shouldn't have `panic!()` in capsules. Change this to a return code that signals the error to the app.

bradjc commented on capsules/src/analog_comparator.rs:
```diff
@@ -0,0 +1,156 @@
+    // Do a single comparison on a channel
+    fn comparison(&self, channel: usize) -> ReturnCode {
+        if channel >= self.channels.len() {
+            panic!("Please select a channel which exists on the current board/chip");
```
here too

bradjc commented on capsules/src/analog_comparator.rs:
```diff
@@ -0,0 +1,156 @@
+    // Stop comparing on a channel
+    fn stop_comparing(&self, channel: usize) -> ReturnCode {
+        if channel >= self.channels.len() {
+            panic!("Please select a channel which exists on the current board/chip");
```
here too

bradjc commented on capsules/src/analog_comparator.rs:
```diff
@@ -0,0 +1,156 @@
+        // Callback to userland
+        self.callback
+            .get()
+            .map_or_else(|| false, |mut cb| cb.schedule(channel, 0, 0));
```
If the array of channels passed to `new()` is re-ordered, does this still work? I'm not sure how the index is associated with the channel.

bradjc commented on kernel/src/hil/analog_comparator.rs:
```diff
@@ -0,0 +1,30 @@
+
+pub trait Client {
+    /// Called when an interrupt occurs.
+    fn fired(&self, usize);
```
More documentation needed, in particular how this unnamed usize maps to the `Channel` above.

dverhaert replied:
It doesn't. Currently, by checking the interrupt status register the used channel is determined, which is done by some simple `if` statements: I don't believe it's worth it to add to the complexity by adding channels here: I've given it a shot, but it got too complex. I noticed the ADC also immediately convert the Channel to a `usize` as well in the chip-specific code/capsule; it only uses the enum to make sure a valid channel is addressed by checking ` self.channels.len()`

If you do believe there is a better way by keeping Channel in and keeping things clear let me know: in that case, we should change it in the ADC as well (basically, I've taken the same approach as there).

bradjc replied:
I don't have an issue with using the `usize` here, just that the comment needs to explain what it is.

bradjc replied:
Still need comment here.

dverhaert replied:
Oops, it appears I added in the capsule not the hil. Done now!

bradjc commented on boards/hail/src/main.rs:
```diff
@@ -127,6 +133,8 @@ unsafe fn set_pin_primary_functions() {
     PA[06].configure(Some(A)); // DAC
     PA[07].configure(None); //... WKP - Wakeup
+                            // PA[06].configure(Some(E)); // ACAN0 - ACIFC
+                            // PA[07].configure(Some(E)); // ACAP0 - ACIFC
```
These comments are good, but I don't think we should have a syscall exposed that no apps can actually used (no assigned pins).

bradjc replied:
I think the example usage in the capsule is enough. Also, if we can figure out shared components then it would be super easy to add in if a user wants to.

dverhaert replied:
Done and done! Let me know if you have any other feedback

bradjc commented on boards/hail/src/main.rs:
```diff
@@ -83,6 +83,10 @@ struct Hail {
+    analog_comparator: &'static capsules::analog_comparator::AnalogComparator<
+        'static,
+        sam4l::acifc::Acifc<'static>,
+    >,
```
:) This is the opposite, please put the pin comments back and remove the code that isn't really connected to anything.

dverhaert replied:
Do you want me to remove the referneces in the struct/impl as well?

bradjc commented:
After 4+ months I think this is time for last call. 

This is a full-stack add for analog comparators, including a HIL interface, implementation for SAM4L, userland interface and capsule, and imix implementation. It also includes the syscall doc for the new syscall interface.

Here is the relevant HIL:

https://github.com/tock/tock/blob/6c510e5672cb8c6f6af44b90629c114d66c4ea4f/kernel/src/hil/analog_comparator.rs#L8-L31

bradjc commented:
@bors r+

bradjc commented:
bors r+
